pydantic/networks.py

```diff
@@ -96,7 +96,31 @@ def __hash__(self) -> int:
     @property
     def defined_constraints(self) -> dict[str, Any]:
         """Fetch a key / value mapping of constraints to values that are not None. Used for core schema updates."""
-        return {field.name: getattr(self, field.name) for field in fields(self)}
+        return {field.name: value for field in fields(self) if (value := getattr(self, field.name)) is not None}
+
+    def __get_pydantic_core_schema__(self, source: Any, handler: GetCoreSchemaHandler) -> core_schema.CoreSchema:
+        schema = handler(source)
+
+        if schema['type'] == 'function-wrap':
+            # for function-wrap schemas, url constraints is applied to the inner schema
+            inner = schema['schema']
+
+            if annotated_type := inner['type'] not in ('url', 'multi-host-url'):
+                raise PydanticUserError(
+                    f"'UrlConstraints' cannot annotate '{annotated_type}'.", code='invalid-annotated-type'
+                )
+            for constraint_key, constraint_value in self.defined_constraints.items():
+                inner[constraint_key] = constraint_value
+            schema['schema'] = inner
+        else:
+            if annotated_type := schema['type'] not in ('url', 'multi-host-url'):
+                raise PydanticUserError(
```

Comment:
Hmm, that validator would be pretty complex, there's a lot involved in `UrlConstraints`...

I don't disagree with the premise, but I don't think this is the place to implement such a change

Reply:
Is there a `url_constraints_schema` in core? I don't think so?

Perhaps this can just go along with https://github.com/pydantic/pydantic/issues/10783...

Reply:
There isn't one but I think it would make sense in a world where constraints are wrapping validators which core can occasionally optimize.

Reply:
Ah yes, along the core optimizing constraint application line of thought